File: articles/CppClasses/Rectangles.cpp
// Rectangle.cpp :
//


#include <iostream>
#include <format>
#include <numbers>
using std::numbers::pi;
using namespace std;

//Base Rectangle class
class Rectangle {
  
    // The dimensions can be accessed by child classes
    protected:
        float width, height;

    public:
        //copy arguments into width and height
        Rectangle(float w, float h) : width{ w }, height{ h } {
       
        }
       /* Rectangle(float w, float h) {
            width = w;
            height = h;
        }*/
        float area() {
            return height * width;
        }
};

// Derived Square class uses area method in parent class
class Square : public Rectangle {

    public:
        Square(float w) : Rectangle(w, w) {}
};


// Derived Semicircle overrides area method
class Semicircle : public Square {
    public:
        Semicircle(float radius):Square(radius) {}
    public:
        float area() {
            return (pi * width * width) / 2;
        }

};

int main() {
    Rectangle rect(5, 4);
    //Rectangle rect = Rectangle(5, 4);
    cout << "Rectangle: " << rect.area() << endl;

    Square sq(6);
    cout << "Square: " << sq.area() << endl;

    Semicircle semi(3);
    //std::cout << "Semicircle: " << semi.area() << std::endl;
    cout << format("Semicircle: {:3.2f}", semi.area());

}

File: articles/CppClasses/Rectangles.py
import math
"""Simple inheritance demonstration using 
   Rectangle, Square and Semicircle"""

# The Rectangle base class
class Rectangle:
    def __init__(self, width, height):
        self.width = width      # save the dimensions
        self.height = height
    def area(self):             # compute the area
        return self.height * self.width

#  a Square is derived from Rectangle and has only one dimension
class Square(Rectangle):
     def __init__(self, side):
         super().__init__( side, side)  #pass both dimensions to parent class

# Semicircle is derived from Square and needs only the radius
class SemiCircle(Square):
     def __init__(self, radius):
         super().__init__( radius)

    # area computation is different in this derived class
     def area(self):
         return (math.pi * self.width * self.width)/2

def main():
     rect= Rectangle(5, 4)
     print("Rectangle,", rect.area())

     sq = Square(6)
     print("Square,", sq.area())

     semi = SemiCircle(3)
     print(f'Semicircle, {semi.area():5.2f} ')

     ###  Here we go  ####


if __name__ == "__main__":
    main()